raster-foundry/granary
# Add user model and only allow interaction with owned objects

## Overview

Brief description of what this PR does, and why it is needed.

### Checklist

- [x] Any new migrations have been audited to make sure they won't kill the application while being applied: tokens will get truncated so we'll need to create new ones, but that's pretty easy and a manual step that was going to follow this anyway

### Notes

This _should_ be what we want I think -- it's now the case that every non-results endpoint on tasks and executions requires either that auth is disabled or that a token is available matching a UUID a user provides in the Authorization header. 

There's one sort of weird effect of this strategy, which is that if you disable auth after having auth enabled and creating some records, create some executions for tasks, then re-enable auth, you won't see the tasks that were created when auth was disabled (since they won't have the `owner` field set). I think that's enough of a corner-case that we don't need to care, but I want to call it out anyway.

## Testing Instructions

- `api/assembly` in `sbt`
- set `AUTH_ENABLED` to `true` in `docker-compose.yml` in the `api` service
- insert a token into your database: `insert into tokens (id, email, user_id) values (uuid_generate_v4(), 'dev@rasterfoundry.com', uuid_generate_v4()) returning id;`
- set your JWT in bash: `export JWT_AUTH_TOKEN='<that id>'`
- confirm you're allowed to list things: `http --auth-type=jwt :8080/api/tasks`, ``http --auth-type=jwt :8080/api/executions`
- create a task by saving the example task from [the docs](https://raster-foundry.github.io/granary/docs/introduction#whats-a-task) to `task.json`, then `cat task.json | http --auth-type=jwt :8080/api/tasks`
- list tasks again and confirm you can see it
- set your JWT to something else and confirm that you can't list things
- unset your JWT and confirm that you can't list things
- create another jwt, set JWT_AUTH_TOKEN to its value, and confirm that you can't see the task you created
- disable auth and make some list requests -- you should see all of the unowned tasks and executions but not the ones that are owned

Closes #260
